File: src/etch/bindings/csharp/transport/DefaultDeliveryService2.cs
using System;
using System.Text;
using System.Collections.Generic;
using Etch.Msg;
using Etch.Support;

namespace Etch.Transport
{
    abstract public class DefaultDeliveryService<T> : AbstractDeliveryService, MessageHandler<MailboxMessageSource> where T : StubBase
    {

        private MailboxMessageSource src;
    	
	    public DefaultDeliveryService()
	    {
		    // Nothing to do.
	    }
    
        /// <summary>
        /// Constructs the MyDeliveryService
        /// </summary>
        /// <param name="src"></param>
	    public DefaultDeliveryService( MailboxMessageSource src )
	    {
		    this.src = src;
	    }

        abstract protected T NewStub( MailboxMessageSource src );

        /*
        /// <summary>
        /// 
        /// </summary>
        /// <param name="src"></param>
        /// Exception:
        ///     throws Exception 
        public void Down(MailboxMessageSource src) 
	    {
            Console.WriteLine( "messageHandlerDown", "who", this, "src", src );
		    //Log.Report( "messageHandlerDown", "who", this, "src", src );

            if ( stub != null )
            {
                stub.Message( src, null, null );
                stub = null;
            }
    	}

        /// <summary>
        /// 
        /// </summary>
        /// <param name="src"></param>
        /// Exception:
        ///     throws Exception 
        public void Up(MailboxMessageSource src) 
	    {
            Console.WriteLine( "messageHandlerUp", "who", this, "src", src );
		    //Log.Report( "messageHandlerUp", "who", this, "src", src );
		    stub = NewStub( src );
    	}
        */



        public static DefaultDeliveryService<T> SetUp( DefaultDeliveryService<T> dds,
            DataSource c, ValueFactory vf )
        {
            MailboxManager mgr = new MailboxManager( dds, 0 );

            Messagizer m = new Messagizer( mgr, vf );
            mgr.SetSource( m );

            Packetizer p = new Packetizer( m, Packetizer.DEFAULT_MAX_PKT_SIZE );
            m.SetSource( p );
            p.SetSource( c );

            dds.SetSource( mgr );
            c.SetHandler( p );

            return dds;
        }

        public void SessionNotify( Object eventObj )
        {
            if ( eventObj.Equals( SessionConsts.UP ) )
                AddStub( NewStub( GetSource() ) );

            base.SessionNotify( eventObj );

            if ( eventObj.Equals( SessionConsts.DOWN ) )
                RemoveStub();
        }
    }
}

File: src/etch/bindings/java/transport/DefaultDeliveryService2.java
/* $Id$
 *
 * Created by wert on Mar 26, 2007.
 *
 * Copyright (c) 2007 Cisco Systems, Inc. All rights reserved.
 */

package etch.bindings.java.transport;


import metreos.core.io.DataSource;
import metreos.core.io.Packetizer;
import etch.bindings.java.msg.ValueFactory;
import etch.bindings.java.support.StubBase;

/**
 * Another variation on DeliveryService.
 * @param <T> the type of the stub.
 */
abstract public class DefaultDeliveryService2<T extends StubBase>
	extends AbstractDeliveryService
	implements MessageHandler<MailboxMessageSource> 

{
	/**
	 * Constructs the DefaultDeliveryService2.
	 */
	public DefaultDeliveryService2()
	{
		// Nothing to do.
	}
	
	/**
	 * @param src
	 * @return new stub
	 */
	abstract protected T newStub( MailboxMessageSource src );
	
//	public void down(MailboxMessageSource src) throws Exception 
//	{
//		
//		Log.report( "messageHandlerDown", "who", this, "src", src );
//		if (stub != null)
//		{
//			stub.message( null, null );
//			stub = null;
//		}
//		
//	}

//	public void up(MailboxMessageSource src) throws Exception 
//	{
//		
//		Log.report( "messageHandlerUp", "who", this, "src", src );
//		stub = newStub( src );
//		
//	}
	
	/**
	 * @param <T>
	 * @param dds
	 * @param c
	 * @param vf
	 * @return an instance of delivery service to use as a message handler.
	 */
	public static <T extends StubBase> DefaultDeliveryService2<T>
		setUp( DefaultDeliveryService2<T> dds, DataSource c,  ValueFactory vf)
	{
		MailboxManager mgr = new MailboxManager( dds, 0 );
		
		Messagizer m = new Messagizer( mgr, vf );
		mgr.setSource( m );
		
		Packetizer p = new Packetizer( m, Packetizer.DEFAULT_MAX_PKT_SIZE );
		m.setSource( p );
		p.setSource( c );
		
		dds.setSource(mgr);
		c.setHandler(p);
		
		return dds;
	}
	
	@Override
	public void sessionNotify( Object event )
	{
		if (event.equals( Session.UP ))
		{
			addStub( newStub( getSource() ) );
		}
		
		super.sessionNotify( event );
		
		if (event.equals( Session.DOWN ))
		{
			removeStub();
		}
	}
}
